### ISSUE TITLE
Adding support for Kotlin

### ISSUE THREAD

**Issue by AshGreh** [powerbot/RSBot-API#2149]
Hey,
after working with java for quite some time, I think it's time we consider moving to or at least supporting kotlin for the API and scripts. This is mainly because kotlin is able to provide a lot of functionality that java isn't.
The following are the reasons as to why I would love to include it,

- Default variables
- Null safety
- Interoperability with java(to a good extent)
- Better function types
- Smart casts
- Simpler class variations
- Data classs
- Getters and Setters (Fields are properties)
- No ternary operators
- Extension of classes without decorators
- Better in loops
- val and var for mutability control
- Constructor in class declaration
- Native delegation
- Ranges and progression(like in python)
- Operator overloading
- Asynchronous Programming through Coroutines(launch and forget for the most part, basically a lighter ideology than threads)
These are just a few of the many pros.

Now for the cons,

- A bit of learning curve for many
- May need re translation in case kotlin support drops(mostly automatable though)

Please let me know what your thoughts are(also, will add any other cons you guys may have to the list if you mention it here)

**Comment by AshGreh** [comment]
An interesting point was brought up by @Mkerian10 on discord. I'm adding this here so as to keep a track of everything. He wanted to add rewrite time to the con for kotlin. I wouldn't say it is, as, I want to break the complete adoption of kotlin into two phases,

- Kotlin support for script writers(on sdn).
- API rewrite in Kotlin

My main goal is to get the first one rolling ASAP and do the second one somewhere down the lifecycle of the bot.

I'll still add this to the con list of Kotlin.

**Comment by Strikegently** [comment]
There is a huge difference between "adding support for Kotlin" and "rewriting everything to Kotlin."
The latter is absolutely not going to happen.

**Comment by AshGreh** [comment]
Is there a reason why the latter can't happen?

**Comment by Paris** [comment]
The primary RSBot codebase is written in Java with over 10 years of incremental improvements. This makes it a robust time-tested software though not without its faults. Rewriting it from scratch in Kotlin is effectively a time wasting task. Many of the cool syntax features like coroutines can already be done in Java 8 using [fork join pools](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html), although it is a bit more verbose when you're familiar with it it becomes straightforward. [NIO allows async](https://docs.oracle.com/javase/8/docs/api/java/nio/package-summary.html) operations too so, just slightly more verbose syntax but you're given more control.

When we migrate to Java 11 soon we'll have even more functionality that Kotlin provides.

However, since Kotlin compiles to Java bytecode as far as I can remember it could be an option to allow script users to use it. I can adapt the SDN compiler to recognise Kotlin code then bundle them all up at the post-compilation bytecode stage. Since Kotlin is supposed to be fully compatible with Java this might not even be a challenging task, but it would take a bit of time and testing. Therefore I'd like to see how many scripters support this idea and would like it implemented.

**Comment by Strikegently** [comment]
With the client and API being Java, it means we can grab almost any developer and they can contribute. 
If we rewrote everything to Kotlin then anyone that wanted to contribute to the client, the API, or make a script would have to learn Kotlin. 

Like I said: there is a huge difference between "adding support for Kotlin" and "rewrite everything to Kotlin." If @Paris is open to adding support for compiling Kotlin scripts on the SDN, then I'm fine with that.

**Comment by Paris** [comment]
I wholeheartedly agree @Strikegently, I wouldn't spend the time rewriting the bot and API which is a very mature codebase that already uses null safety, multithreading and async NIO. We collect bugs in live releases with [Sentry](https://sentry.io/welcome/) and there's virtually zero of those issues any more. So rewriting it is a pointless exercise that would involve a massive rewrite for little gain.

But like you said, an additional language option for script writers may be a viable option since Kotlin is 100% compatible with Java. In theory it would just take a few extra steps to compile Kotlin source to Java bytecode then let the SDN process the rest as normal.

**Comment by constt** [comment]
I think what's been said here so far is all valid, to try and break it down:

- Kotlin is a great language with many great features, it's also becoming more popular amongst java developers and in the RS botting scene.
- The RSBot API is very mature and battle tested, it would be a terrible idea to do a rewrite especially because kotlin plays so well with Java this isn't required.
- Java has a decent periodic release process now and is implementing new features at a decent rate. Many of the nice features kotlin has, Java also does now or is introducing. By continually moving to the latest JDK we can take advantage of these.

All being said I have the following proposal:

- Update SDN to allow kotlin scripts
- Annotate nullable methods/fields on the API so that you get the full power of the kotlin type nullable type checking (Need to investigate this more but shouldn't be too much work).
- Create some tutorials/resources for scripting in kotlin

I think this would give us the best of both worlds as the API would continue to be develop in Java but with good support for Kotlin scripts/plugins. All future additions to the API should be annotated correctly to ensure best compatibility with kotlin.

**Comment by k0na** [comment]
While it wouldn't hurt to support Kotlin, I see SO MANY other things where the time would be more wisely spent (e.g. https://www.powerbot.org/community/topic/1353455-v8-transition-plan/ or a large chunk of the open issues dating back to 2015).

**Comment by ChrisBPB** [comment]
As kona says I think our time needs to be spent adding important features developers need, and then I think UX is currently quite poor, adding kotlin support just seems like something we would consider when we run out of things to do. I don’t believe it makes our product better, or more alluring

**Comment by Paris** [comment]
Yes this is a definitely low priority after Java 11, dark theme and some new hooks.

Adding Kotlin to the SDN is actually relatively easy and would take a day or two's work including testing, but I don't want to get sidetracked as one typically does.

This thread was mainly to gather feedback and so far I've decided that we will soon support Kotlin for scripts only. As it's 100% compatible with Java there should be no issues with the RSBot and its API still being in Java.

I can justify the time of work on it because Kotlin is a trendy new language and it may encourage new developers, or existing ones to convert their code so it performs better.

**Comment by AshGreh** [comment]
@Paris I disagree with achieving coroutines through fork join. A good portion of the suspend system would require a mini framework. I do get your point though. The amount of yield vs the effort is negative on our end. I was considering more of a change as needed support for kotlin. After our conversation on discord, I went through j14-15 spec, doesn't seem like they have answered much of the features I have written up in the pros, but I do see your point.

@Strikegently Considering how most new developers prefer python esque syntax over what java is, I have to disagree with you on that one.

@constt Considering how the j15 spec is at the moment, I wouldn't rely on keeping my hopes up with java. I do agree with you to a certain extent, but realistically, the same robustness, if not more can be achieved with kotlin while writing less than what we have for v8.

@k0na I somewhat agree with you, but, a few of the open issues would benefit from the language features I mentioned at the top.

@ChrisBPB kotlin is a better practice overall which will enable me and few other developers to do much more in much less.

Again, i'm just spit balling ideas here. I just want as many QoL features added as we can feasibly include.

**Comment by Paris** [comment]
I decided to support Kotlin, it shouldn't require too much work and supporting a new language does have its benefits. We will need to include the Kotlin standard library but this should be straightforward.
